File: app/api/routers/finance.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from fastapi import APIRouter, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.deps import DBSession
from app.infra.models import FinanceORM, FinanceStatus, WppSendStatus
from app.schemas.finance import FinanceCreate, FinanceUpdate, FinancePay, FinanceOut

router = APIRouter()


@router.post("", response_model=FinanceOut, status_code=201)
def create_finance(payload: FinanceCreate, db: Session = DBSession):
    try:
        status = FinanceStatus(payload.status)
    except Exception:
        raise HTTPException(status_code=400, detail="status inválido (PENDING|PAID|CANCELED).")

    row = FinanceORM(
        company=payload.company.strip(),
        amount=payload.amount,
        due_date=payload.due_date,
        status=status,
        description=payload.description,
        notes=payload.notes,
        wpp_status=WppSendStatus.PENDING,
        wpp_tries=0,
        wpp_last_error=None,
        wpp_sent_at=None,
        wpp_next_retry_at=None,
    )
    db.add(row)
    db.flush()
    return row


@router.get("", response_model=list[FinanceOut])
def list_finance(
    db: Session = DBSession,
    status: Optional[str] = Query(default=None, description="PENDING|PAID|CANCELED"),
    company: Optional[str] = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
):
    stmt = select(FinanceORM).order_by(FinanceORM.due_date.asc(), FinanceORM.id.asc())

    if status:
        try:
            st = FinanceStatus(status)
        except Exception:
            raise HTTPException(status_code=400, detail="status inválido (PENDING|PAID|CANCELED).")
        stmt = stmt.where(FinanceORM.status == st)

    if company:
        stmt = stmt.where(FinanceORM.company.ilike(f"%{company}%"))

    stmt = stmt.limit(limit).offset(offset)
    return db.execute(stmt).scalars().all()


@router.get("/{finance_id}", response_model=FinanceOut)
def get_finance(finance_id: int, db: Session = DBSession):
    row = db.get(FinanceORM, finance_id)
    if not row:
        raise HTTPException(status_code=404, detail="Finance não encontrado.")
    return row


@router.put("/{finance_id}", response_model=FinanceOut)
def update_finance(finance_id: int, payload: FinanceUpdate, db: Session = DBSession):
    row = db.get(FinanceORM, finance_id)
    if not row:
        raise HTTPException(status_code=404, detail="Finance não encontrado.")

    if payload.company is not None:
        row.company = payload.company.strip()
    if payload.amount is not None:
        row.amount = payload.amount
    if payload.due_date is not None:
        row.due_date = payload.due_date
    if payload.description is not None:
        row.description = payload.description
    if payload.notes is not None:
        row.notes = payload.notes

    if payload.status is not None:
        try:
            row.status = FinanceStatus(payload.status)
        except Exception:
            raise HTTPException(status_code=400, detail="status inválido (PENDING|PAID|CANCELED).")

    db.flush()
    return row


@router.post("/{finance_id}/pay", response_model=FinanceOut)
def pay_finance(finance_id: int, payload: FinancePay, db: Session = DBSession):
    row = db.get(FinanceORM, finance_id)
    if not row:
        raise HTTPException(status_code=404, detail="Finance não encontrado.")

    # marca como pago
    row.status = FinanceStatus.PAID

    row.wpp_status = WppSendStatus.SENT
    row.wpp_sent_at = payload.paid_at or datetime.utcnow()
    row.wpp_next_retry_at = None
    row.wpp_last_error = None

    db.flush()
    return row

File: wi-front/src/services/finance.ts
import { api } from "./api";

export type FinanceStatus = "PENDING" | "PAID" | "CANCELED";
export type WppSendStatus = "PENDING" | "SENDING" | "SENT" | "FAILED";

export type Finance = {
  id: number;
  company: string;
  amount: string; // decimal vem como string
  due_date: string;
  status: FinanceStatus;

  description?: string | null;
  notes?: string | null;

  wpp_status?: WppSendStatus | null;
  wpp_tries?: number | null;
  wpp_last_error?: string | null;
  wpp_sent_at?: string | null;
  wpp_next_retry_at?: string | null;

  created_at?: string | null;
};

export type FinanceCreate = {
  company: string;
  amount: number | string; // aceita number no front
  due_date: string;
  status?: FinanceStatus;
  description?: string;
  notes?: string;
};

export type FinanceUpdate = Partial<FinanceCreate>;

export async function listFinance(): Promise<Finance[]> {
  const { data } = await api.get<Finance[]>("/finance");
  return data;
}

export async function getFinance(financeId: number): Promise<Finance> {
  const { data } = await api.get<Finance>(`/finance/${financeId}`);
  return data;
}

export async function createFinance(payload: FinanceCreate): Promise<Finance> {
  // garante amount em string
  const body = {
    ...payload,
    amount: typeof payload.amount === "number" ? payload.amount.toFixed(2) : payload.amount,
    status: payload.status ?? "PENDING",
  };

  const { data } = await api.post<Finance>("/finance", body, {
    headers: { "Content-Type": "application/json" },
  });
  return data;
}

export async function updateFinance(financeId: number, payload: FinanceUpdate): Promise<Finance> {
  const body: any = { ...payload };

  if (body.amount != null) {
    body.amount = typeof body.amount === "number" ? body.amount.toFixed(2) : body.amount;
  }

  const { data } = await api.put<Finance>(`/finance/${financeId}`, body, {
    headers: { "Content-Type": "application/json" },
  });
  return data;
}

export async function payFinance(financeId: number): Promise<Finance> {
  const { data } = await api.post<Finance>(`/finance/${financeId}/pay`, {}, {
    headers: { "Content-Type": "application/json" },
  });
  return data;
}
